### PR TITLE
Use Git extension API

### PR DESCRIPTION
**Pull request by joaomoreno**
Related to https://github.com/Microsoft/vscode/issues/31103

This is ongoing work. The API exposure will be added on the git extension as we go along. Rough sketch:

- [x] Adopt `exec` and `spawn`
- [x] Adopt `Repository`

### CODE REVIEW

**Comment by joaomoreno** [comment]
@rebornix Are we using these additional arguments for a specific reason?

https://github.com/Microsoft/vscode-pull-request-github/blob/c0e6044e675a074633c9e2dcc22d1da708b35c0d/src/common/git.ts#L38

**Comment by joaomoreno** [comment]
@rebornix @RMacfarlane 

I've gone through all git usage of this extension and exported just enough git extension API to get it working. Here's the current git extension API typings file:

https://github.com/Microsoft/vscode/blob/master/extensions/git/src/api/git.d.ts


Haven't tested it yet, will have to do it next week. Feel free to start reviewing and properly smoke testing the extension. Note that since this is **really** fresh, you'd going to have to use the extension against latest `master` dev, since we still don't have an insider build in which the git extension actually exposes this much API.

**Comment by joaomoreno** [comment]
Merged `master` back into PR. Did some testing. Found a couple issues, fixed. Found missing implementation on the git extension side, fixed in Microsoft/vscode. Waiting for tomorrow's build to continue.

Need to test for #341, might've broken

**Comment by joaomoreno** [comment]
Merged `master` into this once again. Found another issue in the core git extension, pushed fix for that: https://github.com/Microsoft/vscode/commit/f92afbdc3ced5f7710d9be1c2eed4da76c490652 Other than that, this branch seems to behave correctly from my sanity checking.

@rebornix @RMacfarlane Once we have a new insiders build containing https://github.com/Microsoft/vscode/commit/f92afbdc3ced5f7710d9be1c2eed4da76c490652, please give this PR a smoketest. We're getting closer. :+1:

**Comment by joaomoreno** [comment]
@rebornix @RMacfarlane Let's review this ASAP. I've merged upstream changes once again, there were 9 conflicting files.

File: src/extension.ts [hunk 1]
```diff
@@ -5,69 +5,71 @@
...
-	const repository = new Repository(rootPath);
-	let repositoryInitialized = false;
-	let prManager: PullRequestManager;
 
-	repository.onDidRunGitStatus(async e => {
-		if (repositoryInitialized) {
-			return;
-		}
+	const rootPath = vscode.workspace.rootPath;
+	// const repository = new Repository(rootPath);
```
Comment (RMacfarlane) [inline src/extension.ts]: nit: remove commented out code

File: src/extension.ts [hunk 2]
```diff
@@ -5,69 +5,71 @@
...
-		prManager = new PullRequestManager(configuration, repository, telemetry);
-		const reviewManager = new ReviewManager(context, configuration, repository, prManager, telemetry);
-		registerCommands(context, prManager, reviewManager, telemetry);
-		telemetry.on('startup');
-	});
+	if (repository) {
+		await init(context, repository);
+	} else {
+		const onDidOpenRelevantRepository = filterEvent(api.onDidOpenRepository, r => isDescendant(r.rootUri.fsPath, rootPath));
+		onceEvent(onDidOpenRelevantRepository)(repository => init(context, repository));
```
Comment (RMacfarlane) [inline src/extension.ts]: consider renaming `repository` since it's currently shadowing repository above

File: src/view/reviewManager.ts [hunk 3]
```diff
@@ -73,9 +72,9 @@ export class ReviewManager implements vscode.DecorationProvider {
 				opts.selection = activeTextEditor.selection;
 			}
 
-			vscode.commands.executeCommand('vscode.open', vscode.Uri.file(nodePath.resolve(this._repository.path, params.path)), opts);
+			vscode.commands.executeCommand('vscode.open', vscode.Uri.file(nodePath.resolve(this._repository.rootUri.fsPath, params.path)), opts);
 		}));
-		this._disposables.push(_repository.onDidRunGitStatus(e => {
+		this._disposables.push(_repository.state.onDidChange(e => {
 			// todo, validate state only when state changes.
 			this.updateState();
```
Comment (RMacfarlane) [inline src/view/reviewManager.ts]: we have an existing problem with this that I might write a change for right now (and cause you another merge conflict, sorry), but we really only need to call `updateState` when the repo's HEAD or remotes have changed

Testing these changes made me realize how much of an issue this is :scream

File: src/github/pullRequestOverview.ts [hunk 4]
```diff
@@ -250,23 +249,17 @@ export class PullRequestOverviewPanel {
...
 
-				if (`refs/heads/${branch}` === branchFullName) {
-					await this._pullRequestManager.checkout(branch);
-				} else {
-					await vscode.commands.executeCommand('git.checkout');
-				}
+			if (branch === branchObj.name) {
+				await this._pullRequestManager.repository.checkout(branch);
+			} else {
+				await vscode.commands.executeCommand('git.checkout')
```
Comment (RMacfarlane) [inline src/github/pullRequestOverview.ts]: missing ;

File: src/view/treeNodes/pullRequestNode.ts [hunk 5]
```diff
@@ -402,13 +403,13 @@ export class PRNode extends TreeNode {
 					return right.join('\n');
 				}
 			} else {
-				if (params.base) {
-					const originalFileName = fileChange.status === GitChangeType.RENAME ? fileChange.previousFileName : fileChange.fileName;
-					return getFileContent(this.repository.path, params.commit, originalFileName);
+				const originalFileName = fileChange.status === GitChangeType.RENAME ? fileChange.previousFileName : fileChange.fileName;
+				const originalFilePath = path.join(this.repository.rootUri.fsPath, originalFileName);
+				const originalContent = await this.repository.show(params.commit, originalFilePath);
```
Comment (rebornix) [inline src/view/treeNodes/pullRequestNode.ts]: It seems that sometimes the result from the api ends with an additional line break. For example in https://github.com/auchenberg/pullrequest-demo/pull/6/files , `app.js`'s last line is not empty.

File: src/typings/git.d.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+export interface Repository {
+
+	readonly rootUri: Uri;
+	readonly inputBox: InputBox;
+	readonly state: RepositoryState;
+
+	getConfigs(): Promise<{ key: string; value: string; }[]>;
+	getConfig(key: string): Promise<string>;
```
Comment (rebornix) [inline src/typings/git.d.ts]: `getConfig` result contains line breaks as well.

File: src/typings/git.d.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+	diffIndexWithHEAD(path: string): Promise<string>;
+	diffIndexWith(ref: string, path: string): Promise<string>;
+	diffBlobs(object1: string, object2: string): Promise<string>;
+	diffBetween(ref1: string, ref2: string, path: string): Promise<string>;
+
+	hashObject(data: string): Promise<string>;
+
+	createBranch(name: string, checkout: boolean, ref?: string): Promise<void>;
+	deleteBranch(name: string): Promise<void>;
+	getBranch(name: string): Promise<Branch>;
```
Comment (rebornix) [inline src/typings/git.d.ts]: Run into one exception with `getBranch`. Try to checkout https://github.com/auchenberg/pullrequest-demo/pull/12
```
"fatal: ambiguous argument 'misolori/update': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
"
```

**Comment by rebornix** [comment]
@joaomoreno you can leave the conflicts resolving work to me, and we can do that when the branch before we merge.

With git's API everything becomes way simpler than before, and the change looks great. My only concerns right now are

* at the very beginning we run git commands directly in command line and it doesn't throw, now the git API may throw when running into errors. This PR already `try/catch` in most cases, but we missed some, like https://github.com/Microsoft/vscode-pull-request-github/pull/254/files#diff-ac1a52fba835c4784937e91b26b6ba31R31 . If you don't mind I'd like to push fixes for them directly to your branch.
* Normalization of git API result. We used to trim the line ending of the return result from command line. It seems we don't do that in Git extension (see comments inline), do you think that's what we can have?

Other parts look good to me, pretty solid.

File: src/common/diffHunk.ts [hunk 8]
```diff
@@ -264,17 +264,21 @@ export async function parseDiff(reviews: any[], repository: Repository, parentCo
...
-		let originalFileExist;
+		let originalFileExist = false;
+
 		switch (gitChangeType) {
 			case GitChangeType.DELETE:
 			case GitChangeType.MODIFY:
-				originalFileExist = await repository.checkFileExistence(parentCommit, review.filename);
+				try {
+					await repository.getObjectDetails(parentCommit, review.filename);
+					originalFileExist = true;
```
[the remark this answers is not in the document]
Comment (joaomoreno) [reply src/common/diffHunk.ts]: There used to be the `checkFileExistence` method, which returned a `boolean`.

Now, we use `getObjectDetails` which returns information about that object or throws if the object does not exist.

File: src/view/treeNodes/pullRequestNode.ts (hunk 5 above)
Comment (joaomoreno) [reply src/view/treeNodes/pullRequestNode.ts]: I can't repro that one. When I open the diff in the PR extension, I get the same diff as in GitHub:

image

File: src/typings/git.d.ts (hunk 7 above)
Comment (joaomoreno) [reply src/typings/git.d.ts]: Good catch. There were also other calls to getBranch which were just not aware of the throw when the branch didn't exist. Fixed those as well.

**Comment by joaomoreno** [comment]
🤦‍♂️ For some reason I didn't have tslint in the machine I did the work. Sorry about that guys.

Addressed feedback.

> If you don't mind I'd like to push fixes for them directly to your branch.

Absolutely, go ahead!

> Normalization of git API result. We used to trim the line ending of the return result from command line. It seems we don't do that in Git extension (see comments inline), do you think that's what we can have?

Fixed it for `getConfig`. 👍 Not sure about for `show`, commented inline.

Found another issue with `getBranch('@{-1}')`, pushed fix in vscode for it: https://github.com/Microsoft/vscode/commit/2c0167ef9906e7ee765f99e3b1244de2e56bd4e8

File: src/view/reviewManager.ts (hunk 3 above)
Comment (joaomoreno) [reply src/view/reviewManager.ts]: > but we really only need to call updateState when the repo's HEAD or remotes have changed

Yeah we can always filter out the events which should not reflect a state update.

File: src/view/treeNodes/pullRequestNode.ts (hunk 5 above)
Comment (rebornix) [reply src/view/treeNodes/pullRequestNode.ts]: You can probably see this once you have line numbers turned on. Screenshots from latest binary

image

GitHub's diff

image

We get a line break at the end so we have line 59 in the diff view.

**Comment by rebornix** [comment]
@joaomoreno played with latest code, the only issue I ran into is https://github.com/Microsoft/vscode-pull-request-github/pull/254#discussion_r217894206 mentioned above. Once that one is addressed, IMO we can merge it.

File: src/view/treeNodes/pullRequestNode.ts (hunk 5 above)
Comment (rebornix) [reply src/view/treeNodes/pullRequestNode.ts]: @joaomoreno thanks for the detailed investigation, I verified that the diff view on GitHub doesn't  show the empty line always somehow. @shana do you have insights on this?

**Comment by rebornix** [comment]
@joaomoreno add protection for NPEs, it's safe now. @RMacfarlane can take another look then we can merge this PR.

**Comment by joaomoreno** [comment]
Yeah! :fireworks: Thanks for the help!

**Pull request merged**